pkg/registry/core/service/strategy.go

```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
+// any other option.
+func dropTypeDependentFields(newSvc *api.Service, oldSvc *api.Service) {
+	// For now we are only normalizing on updates.  This minimizes potential
+	// confusion since many of the cases we are handling here are pretty niche.
+	if oldSvc == nil {
+		return
+	}
+
+	// If the user is switching to a type that does not need a value (even
+	// "None" counts), AND they did not change the ClusterIP value, we can wipe
+	// it.  It will be deallocated later.
+	//
+	// There are other fields that could also be wiped, but historically were
+	// ignored.  For compat, we leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) && sameClusterIP(oldSvc, newSvc) {
+		newSvc.Spec.ClusterIP = ""
+	}
+
+	// If the user is switching to a type that doesn't use NodePorts AND they
+	// did not change any NodePort values, we can wipe them.  It will be
+	// deallocated later.
+	if needsNodePort(oldSvc) && !needsNodePort(newSvc) && sameNodePorts(oldSvc, newSvc) {
+		for i := range newSvc.Spec.Ports {
+			newSvc.Spec.Ports[i].NodePort = 0
+		}
+	}
+
+	// If the user is switching to a case that doesn't use HealthCheckNodePort AND they
+	// did not change the HealthCheckNodePort value, we can wipe it.  It will
+	// be deallocated later.
+	if needsHCNodePort(oldSvc) && !needsHCNodePort(newSvc) && sameHCNodePort(oldSvc, newSvc) {
+		newSvc.Spec.HealthCheckNodePort = 0
+	}
+
+	// NOTE: there are other fields like `selector` which we could wipe.
+	// Historically we did not wipe them and they are not allocated from
+	// finite pools, so we are (currently) choosing to leave them alone.
+}
+
+func needsClusterIP(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeExternalName {
+		return false
+	}
+	return true
+}
+
+func sameClusterIP(oldSvc, newSvc *api.Service) bool {
+	return oldSvc.Spec.ClusterIP == newSvc.Spec.ClusterIP
+}
+
+func needsNodePort(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		return true
+	}
+	return false
+}
+
+func sameNodePorts(oldSvc, newSvc *api.Service) bool {
+	// Helper to make a unique key so we can tell if a port changed.
+	makeKey := func(p *api.ServicePort) string {
+		return fmt.Sprintf("%s/%d", p.Protocol, p.Port)
+	}
+	// Helper to make a map of port key -> NodePort value.
+	mapNodePorts := func(svc *api.Service) map[string]int32 {
+		out := make(map[string]int32)
+		for i := range svc.Spec.Ports {
+			key := makeKey(&svc.Spec.Ports[i])
+			out[key] = svc.Spec.Ports[i].NodePort
+		}
+		return out
+	}
+
+	oldPorts := mapNodePorts(oldSvc)
+	newPorts := mapNodePorts(newSvc)
+
+	// Every old port must have the same NodePort in the new config or be
+	// cleared to 0 or must be removed.
+	for k, oldVal := range oldPorts {
+		if newVal, found := newPorts[k]; found {
+			if newVal != oldVal && newVal != 0 {
+				return false
+			}
+		}
+	}
+	// Every new port must not have a NodePort.
```

Comment:
That does not make it same right? it confuses the user intent with (change type and add new ports)

Reply:
I'm saying that it is valid to change both things AS LONG AS the user did not specify a NodePort value.  This is closer to the create path, where we will barf if you set NodePort.  If the new state does not need a NodePort (e.g. ExternalName) and the new port does not have a NodePort, the user intent is clear.